### DISCUSSION TITLE
Syntax for select and to_list

### DISCUSSION THREAD

**Discussion by fl-hat** [JoelForamitti/agentpy#22] [Q&A] [answered]
Hi! I've been trying to select a group of agents using agentpy.Space.select, and then from those agents select only those with a specific state. I'm running into some trouble with the syntax for `select` and `to_list`. 

Here's what I've been using:
```
nearby_agents = self.mymodel.space.select(midpoint, search_radius) # Select nearby resources
nearby_agents = ap.AgentIter.to_list(nearby_agents) # Change type from AgentIter to AgentList
nearby_resources = nearby_agents(nearby_agents.state == "resource") # Select only the resources
```

I added the second line (using `to_list`) after I got the error `TypeError: 'AgentIter' object is not callable`. 
I've tried three different ways of calling `to_list` and have gotten three different error messages:

```
nearby_agents = nearby_agents.to_list()
AttributeError: Hive (Obj 102) has no attribute 'to_list'.

ap.to_list(nearby_agents)
AttributeError: module 'agentpy' has no attribute 'to_list'

nearby_agents = ap.AgentIter.to_list(nearby_agents)
AttributeError: type object 'AgentIter' has no attribute 'to_list'
```

I'd really appreciate any advice about the syntax, or better ways of doing what I'm trying to accomplish. Thank you so much!

**Comment by jofmi** [comment, marked as answer]
Hi @fl-hat !

Please make sure to update to the latest version (0.1.2), as `to_list` is a new feature.

I think it should work with the following:

    nearby_agents = self.mymodel.space.select(midpoint, search_radius)
    nearby_agents = nearby_agents.to_list()

Or with just one line:

    nearby_agents = self.mymodel.space.select(midpoint, search_radius).to_list()

To improve performance, you could consider to put your resources in a separate `Space`, so that you don't have to select the resources (i.e. your third line of code) every time. E.g. you could have an `agent_space` and a `resource_space`.

**Reply by fl-hat** [reply]
Hi! Thank you so much for your advice, the line you provided worked! I'm still having trouble selecting just the resources (using `nearby_resources = nearby_agents(nearby_agents.state == "resource")` I get `TypeError: 'AgentList' object is not callable`), and like the idea of putting the resources in a separate `Space`. I was wondering then if ipysimulate would be able to support visualizing agents in both spaces simultaneously? Currently, I'm using:
```
scatterplot = ips.Scatterplot(
    control,
    xy=lambda m: m.space.positions.values(),
    c=lambda m: m.agents.state
)
```
and it's not immediately clear to me how to specify a second `Space` to be overlaid. Thanks!

**Reply by jofmi** [reply]
Hi @fl-hat. The feature to call agent lists directly has been deprecated, so now it is `nearby_resources = nearby_agents.select(nearby_agents.state == "resource")`. If there was a part of the documentation that told you otherwise, please let me know!

Regarding plotting both spaces, you would need to iterate over both. You could do this with itertools:

```
import agentpy as ap
import itertools

def MyModel(ap.Model):
    def setup(self):
        self.agents1 = ap.AgentList(...)
        self.agents2 = ap.AgentList(...)
        self.agents = self.agents1 + self.agents2

        self.space1 = ap.Space(...)
        self.space1.add_agents(self.agents1, ...)
        self.space2 = ap.Space(...)        
        self.space2.add_agents(self.agents1, ...)

...

def both_space_positions(m):
    i1 = m.space1.positions.values()
    i2 = m.space2.positions.values()
    return itertools.chain(i1, i2)

scatterplot = ips.Scatterplot(
    control,
    xy=both_space_positions,
    c=lambda m: m.agents.state
)
```

In the future I hope to offer more convenient tools to do this!
